### DISCUSSION TITLE
Handling Concurrency with API

### DISCUSSION THREAD

**Discussion by Anupam-Panda** [jaredhendrickson13/pfsense-api#364] [General]
Hi Jared @jaredhendrickson13 

One general query. How does this API suite handle concurrency and race conditions? This is especially important with the context of updates happening to a single data file "config.xml" in pfSense. Does the API implement some sort of Lock or semaphore to address the potential of "config.xml" getting corrupted or having inconsistent updates on pfSense?

Not sure if it has been already discussed in this forum or not. Could not able find with search filters.

Thanks and cheers!

**Comment by jaredhendrickson13** [comment]
Hey!

Great question! This is mentioned a bit in the ['Limitations' section of the README](https://github.com/jaredhendrickson13/pfsense-api#limitations) but to go into a little more detail:

Both pfSense and the API package contain configuration locks that helps prevent write conflicts. This does make concurrent API calls relatively safe in most situations. However, this is more of a safety net than a feature; concurrent API calls (for write actions at least) are still discouraged for a few reason:

- Some of the core pfSense functions the API calls may not respect these config locks. This is especially problematic in asynchronous/background functions used by pfSense.
- Many API endpoints/pfSense functions also restart backend services which may not tolerate concurrency.
- Some components of pfSense use an epoch timestamp as object IDs, this effectively limits creation of these objects to 1 per second. The firewall rule 'tracker ID' is a good example of this. 

In general, it's best to run API calls in sequence. In some cases a small delay between API calls may be necessary. If you do plan on leveraging concurrency, I strongly recommend testing thoroughly in a non-production environment to determine expected behavior. Hope this helps.

Thanks!
